File: client/net/client/Com.Bj58.Spat.Gaea.Client/Proxy/Builder/MethodCaller.cs
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
//----------------------------------------------------------------------
//<Copyright company="58.com">
//      Team:SPAT
//      Blog:http://blog.58.com/spat/  
//      Website:http://www.58.com
//</Copyright>
//-----------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Reflection;
using Com.Bj58.Spat.Gaea.Client.Proxy.Component;

namespace Com.Bj58.Spat.Gaea.Client.Proxy.Builder
{
	internal class MethodCaller
	{

        private string _serviceName;
        private string _type;

		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="ProxyFactory">ProxyFactory - Instance</param>
		/// <param name="uri">Server-Proxy uri</param>
		public MethodCaller(string serviceName,string type)
		{
            this._serviceName = serviceName;
            this._type = type;
		}

		/// <summary>
		/// This method wrapps an instance call
		/// </summary>
		/// <param name="methodInfo">The method to call</param>
		/// <param name="arrMethodArgs">The arguments to the method call</param>
		public object DoMethodCall(object[] arrMethodArgs, MethodInfo methodInfo)
		{   
            ParameterInfo[] pis = methodInfo.GetParameters();
            if ((arrMethodArgs.Length != pis.Length))
            {
                throw new ArgumentException("Arguments count not match!");
            }
            var proxy = ServiceProxy.GetProxy(_serviceName);
            Parameter[] paras = new Parameter[arrMethodArgs.Length];
            List<int> outParas = new List<int>();
            for (int i = 0; i < pis.Length; i++)
            {
                var pi = pis[i];
                if (pi.IsOut || pi.ParameterType.IsByRef)
                {
                    paras[i] = new Parameter(arrMethodArgs[i], pi.ParameterType.GetElementType(), Com.Bj58.Spat.Gaea.Client.Proxy.Enum.ParaType.Out);
                    outParas.Add(i);
                }
                else
                {
                    paras[i] = new Parameter(arrMethodArgs[i], pi.ParameterType, Com.Bj58.Spat.Gaea.Client.Proxy.Enum.ParaType.In);
                }
            }
            string methodName = methodInfo.Name;
            var atts = methodInfo.GetCustomAttributes(typeof(Com.Bj58.Spat.Gaea.OperationContractAttribute), true);
            if (atts != null && atts.Length > 0)
            {
                OperationContractAttribute ma = (OperationContractAttribute)atts[0];
                if (!string.IsNullOrEmpty(ma.Name))
                {
                    methodName = ma.Name;
                }
                methodName = "$" + methodName;
            }
            var result = proxy.Invoke(methodInfo.ReturnType, _type, methodName, paras);
            for (int i = 0; i < outParas.Count && i < result.OutPara.Length; i++)
            {
                arrMethodArgs[outParas[i]] = result.OutPara[i];
            }
            return result.Result;
		}
    }
}

File: client/java/client/src/com/bj58/spat/gaea/client/proxy/builder/MethodCaller.java
/*
 *  Copyright Beijing 58 Information Technology Co.,Ltd.
 *
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package com.bj58.spat.gaea.client.proxy.builder;

import java.lang.reflect.Method;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import com.bj58.spat.gaea.client.proxy.ServiceProxy;
import com.bj58.spat.gaea.client.proxy.ServiceProxy.InvokeResult;
import com.bj58.spat.gaea.client.utility.logger.ILog;
import com.bj58.spat.gaea.client.utility.logger.LogFactory;
import com.bj58.spat.gaea.server.contract.annotation.AnnotationUtil;
import com.bj58.spat.gaea.server.contract.annotation.OperationContract;
import com.bj58.spat.gaea.server.contract.entity.Out;

/**
 * MethodCaller
 *
 * @author Service Platform Architecture Team (spat@58.com)
 */
public class MethodCaller {

	private static final ILog logger = LogFactory.getLogger(MethodCaller.class);
	private String serviceName;
    private String lookup;
    

    public MethodCaller(String serviceName, String lookup) {
        this.serviceName = serviceName;
        this.lookup = lookup;
    }

    public Object doMethodCall(Object[] args, Method methodInfo) throws Exception, Throwable {
        Type[] typeAry = methodInfo.getGenericParameterTypes();
        Class<?>[] clsAry = methodInfo.getParameterTypes();
        if (args == null) {
            args = new Object[0];
        }
        if (args.length != typeAry.length) {
            throw new Exception("argument count error!");
        }

        ServiceProxy proxy = ServiceProxy.getProxy(serviceName);
        Parameter[] paras = new Parameter[args.length];
        List<Integer> outParas = new ArrayList<Integer>();

        if (typeAry != null) {
            for (int i = 0; i < typeAry.length; i++) {
                if (args[i] instanceof Out) {
                    paras[i] = new Parameter(args[i], clsAry[i], typeAry[i], ParaType.Out);
                    outParas.add(i);
                } else {
                    paras[i] = new Parameter(args[i], clsAry[i], typeAry[i], ParaType.In);
                }
            }
        }
        Parameter returnPara = new Parameter(null, methodInfo.getReturnType(), methodInfo.getGenericReturnType());
        String methodName = methodInfo.getName();
        OperationContract ann = methodInfo.getAnnotation(OperationContract.class);
        if (ann != null) {
            if (!ann.methodName().equals(AnnotationUtil.DEFAULT_VALUE)) {
                methodName = "$" + ann.methodName();
            }
        }
        InvokeResult result = proxy.invoke(returnPara, lookup, methodName, paras);

        if (result != null && result.getOutPara() != null) {
            for (int i = 0; i < outParas.size() && i < result.getOutPara().length; i++) {
                Object op = args[outParas.get(i)];
                if(op instanceof Out){
                    ((Out)op).setOutPara(result.getOutPara()[i]);
                }
            }
        }
        return result.getResult();
    }
}
